docs/design/proposals/collections.md

```diff
@@ -0,0 +1,604 @@
+time and explicitly defined. This creates a gap between KRO's declarative model
+and real-world infrastructure needs, where the number of resources often depends
+on dynamic factors discovered at runtime - how many availability zones are
+healthy, how many tenants exist, how many namespaces match certain criteria.
+
+Users need the ability to provision resources dynamically based on runtime state
+without sacrificing KRO's core properties: declarative configuration,
+predictable execution, and security guarantees. They need the expressive power
+of loops without the unbounded execution and security risks of traditional
+templating languages.
+
+This proposal introduces declarative resource collections. With collections, one
+resource block can create multiple resources based on data discovered at
+runtime. Collections give users the power of for loops while keeping KRO's
+safety guarantees - execution time is still bounded and behavior is still
+predictable.
+
+## Motivation
+
+### Problem Statement
+
+Kubernetes environments are inherently dynamic, and the number of resources
+needed often depends on runtime configuration rather than design-time decisions.
+KRO's current model requires users to statically define each resource block in
+their ResourceGraphDefinition, creating a fundamental mismatch with real-world
+needs.
+
+The core issue is simple: each resource block in an RGD creates exactly one
+Kubernetes resource. If you need 5 Pods, you write 5 resource blocks. If you
+later need 10, you must modify the RGD to add 5 more blocks.
+
+Consider a batch processing application where users specify how many worker Pods
+to run through their custom resource: `workers: 5`. Each worker needs its own
+Pod with specific environment variables and configurations based on its index.
+With KRO today, you cannot express "create N Pods where N equals spec.workers".
+The same limitation appears in data processing pipelines where each stage might
+need a different number of resources:
+`stages: [{name: "extract", replicas: 3}, {name: "transform", replicas: 5}, {name: "load", replicas: 2}]`.
+
+This forces users to either hardcode a fixed number of resources and live with
+the inflexibility, or wrap KRO with templating tools to generate different
+RGDs - defeating KRO's purpose of moving beyond templating to declarative APIs.
+
+What users really need is the ability to say "for each item in this array,
+create a resource" or "create N resources where N comes from the spec" - dynamic
+resource creation based on runtime values while maintaining KRO's declarative
+approach.
+
+You're right - it's a field, not a construct! Let me fix that:
+
+## Proposal
+
+### Overview
+
+This KREP introduces the `forEach` field for resource definitions, enabling a
+single resource block to create multiple resources based on collections. The
+forEach field can iterate over:
+
+- Arrays from the custom resource spec
```

Comment:
Would we ever iterate over maps, or even over structs? Can always be added later, but this struck me as a common pattern in other PLs.

Reply:
Yes, maps are support - i missed highlighting them clearly in the overview section.. The implementation in https://github.com/kubernetes-sigs/kro/pull/632 handles map iteration.

By struct, do you mean iterating over object fields as key-value pairs? or iterating over an array of objects?

Reply:
> By struct, do you mean iterating over object fields as key-value pairs

Yes this. Kind of wild -- I'd probably hold off until we knew the use case better. I could imagine wanting to walk the YAML though.